cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -275,57 +297,129 @@ public List<TopicId> listTopics(NamespaceId namespaceId)
    * shows how messages would be persisted in the topic tables.
    *
    * <pre>
-   * TXN     sequence_id  payload_sequence_id  publish_ts  payload
-   * TXN1    0            0                    ts1          msg1
-   * TXN1    1            0                    ts1          msg2_p0
-   * TXN1    1            1                    ts1          msg2_p1
-   * TXN2    0            0                    ts2          msg3
-   * TXN3    0            0                    ts3          msg4
+   * TXN     sequence_id  payload_sequence_id  publish_ts  payload       payload_remaining_chunks
+   * TXN1    0            0                    ts1          msg1          0
+   * TXN1    1            0                    ts1          msg2_p0       2
+   * TXN1    1            1                    ts1          msg2_p1       1
+   * TXN1    1            2                    ts1          msg2_p2       0
+   * TXN2    0            0                    ts2          msg3          0
+   * TXN3    0            0                    ts3          msg4          0
    * </pre>
    */
   @Nullable
   @Override
   public RollbackDetail publish(StoreRequest request)
       throws TopicNotFoundException, IOException, UnauthorizedException {
     long start = System.currentTimeMillis();
+    CompletableFuture<Void> publishFuture = new CompletableFuture<>();
+    batch.add(new RequestFuture(request, publishFuture));
 
-    batch.add(request);
-    if (!batch.isEmpty()) {
-      int sequenceId = 0;
-      List<Mutation> batchCopy = new ArrayList<>(batch.size());
-      // We need to batch less than fetch limit since we read for publish_ts >= last_message.publish_ts
-      // see fetch for explanation of why we read for publish_ts >= last_message.publish_ts and
-      // not publish_ts > last_message.publish_ts
-      while (!batch.isEmpty()) {
-        StoreRequest headRequest = batch.poll();
-        for (byte[] payload : headRequest) {
-          //TODO: [CDAP-21094] Breakdown messages with larger payload into parts.
-          Mutation mutation = Mutation.newInsertBuilder(getTableName(headRequest.getTopicId()))
-              .set(SEQUENCE_ID_FIELD).to(sequenceId++).set(PAYLOAD_SEQUENCE_ID_FIELD).to(0)
-              .set(PUBLISH_TS_FIELD).to("spanner.commit_timestamp()").set(PAYLOAD_FIELD)
-              .to(ByteArray.copyFrom(payload)).build();
-          batchCopy.add(mutation);
-        }
+    synchronized (clientLock) {
+      writeData(start);
+    }
+    // Wait for the corresponding future to finish.
+    publishFuture.join();
+
+    //TODO: Add a RollbackDetail implementation that throws exceptions if any of the methods is called.
+    return null;
+  }
+
+  private void writeData(long start) {
```

Comment:
Make this a synchronized method, so you don't need the synchronized block above.